compiler/rustc_data_structures/src/obligation_forest/mod.rs

```diff
@@ -117,6 +117,12 @@ pub trait ObligationProcessor {
 }
 
 /// The result type used by `process_obligation`.
+// `repr(C)` to inhibit the niche filling optimization. Otherwise, the `match` appearing
+// in `process_obligations` is significantly slower, which can substantially affect
+// benchmarks like `rustc-perf`'s inflate and keccak. FIXME: When it's feasible to do
+// optimizations after monomorphization, applying `AddNicheCases` or something like it
+// might alleviate this performance issue.
+#[repr(C)]
```

Comment:
https://github.com/rust-lang/rust/pull/94075#issuecomment-1079561695 makes a little confusion to me.
> There were too many valid values in the niche for my MirPass to apply, and the SwitchInt was just being hit repeatedly. 

Do you mean that `process_obligations` is slow when together with the MirPass outside this PR, or we need this `repr(C)` no matter if we have a MirPass?

Reply:
`process_obligations` was slow without the MirPass. The MirPass was an attempt to speed back up some of these matches, but since it could only be applied to non-generic code, it didn't optimize `process_obligations`.

Actually I guess it might be worthwhile for me to do a few more benchmarks and see whether this is still slower without the `repr(C)`. And I should probably edit this comment to not mention `AddNicheCases`, which no longer exists and which will confuse anyone reading this.

I'll get to those tasks in the near future.